source/elastic_solver.py

```diff
@@ -54,13 +48,99 @@ def __init__(self, mesh, boundary_markers, polynomial_degree=1):
         # dimension-dependent variables
         self._null_vector = dlfn.Constant((0., ) * self._space_dim)
 
+        # set elastic law
+        assert isinstance(elastic_law, ElasticLaw)
+        self._elastic_law = elastic_law
+
         # set discretization parameters
         # polynomial degree
         self._p_deg = polynomial_degree
         # quadrature degree
         q_deg = 2 * self._p_deg
         dlfn.parameters["form_compiler"]["quadrature_degree"] = q_deg
+    
+    def _check_boundary_condition_format(self):
+        """
+        Virtual method to check the general format of an arbitrary boundary condition.
+        """
+        raise NotImplementedError("You are calling a purely virtual method.")
+    
+    def set_body_force(self, body_force):
+        """
+        Specifies the body force.
+
+        Parameters
+        ----------
+        body_force : dolfin.Expression, dolfin. Constant
+            The body force.
+        """
+        assert isinstance(body_force, (dlfn.Expression, dlfn.Constant))
+        # check rank of expression
+        assert body_force.value_rank() == 1
+        self._body_force = body_force
+    
+    def set_boundary_conditions(self):
+        """
+        Purely virtual method to set the boundary conditions of the problem.
+        """
+        raise NotImplementedError("You are calling a purely virtual method.")
+    
+    def _setup_boundary_conditions(self):
+        """
+        Purely virtual method to set up the boundary conditions of the problem.
+        """
+        raise NotImplementedError("You are calling a purely virtual method.")
+            
+    def set_dimensionless_numbers(self):
+        """
+        Purely virtual method to update the parameters of the model by creating or modifying class
+        objects.
+        """
+        raise NotImplementedError("You are calling a purely virtual method.")
+    
+    def _setup_function_spaces(self):
+        """
+        Virtual class method setting up function spaces.
+        """
+        raise NotImplementedError("You are calling a purely virtual method.")
+    
+    def _setup_problem(self):
+        """
+        Virtual method to set up solver objects.
+        """
+        raise NotImplementedError("You are calling a purely virtual method.")
```

Comment:
Aren't these most of these setup methods the same for a linear and a non-linear elastic problem? For example, `_check_boundary_condition_format`, `_setup_boundary_conditions`,  `_setup_function_spaces` etc.

Or do you already distinguish between a compressible and incompressible material?